File: example-cs/Puzzle.cs
using System.Collections.Generic;
using System.Linq;

namespace Aoc2021
{
    public class Puzzle : AocPuzzle<int[], int>
    {
        protected override IDictionary<string, int> TestAnswers { get; } = new Dictionary<string, int>
        {
            { "part1", 1379 },
            { "part2", 56154 }
        };

        protected override int[] ParseInput(string[] lines)
        {
            return lines.Select(line => int.Parse(line)).ToArray();
        }

        protected override int RunPart1()
        {
            return _input.Sum();
        }

        protected override int RunPart2()
        {
            return _input.Aggregate((prev, cur) => prev * cur);
        }
    }
}

File: day10/puzzle.py
import statistics
from functools import reduce
from aoc import AocPuzzle


TEST_ANSWERS = {
    "part1": 26397,
    "part2": 288957
}

CHAR_MAP = {
    "(": ")",
    "[": "]",
    "{": "}",
    "<": ">"
}

ILLEGAL_CHAR_POINTS = {
    ")": 3,
    "]": 57,
    "}": 1197,
    ">": 25137
}

COMPLETION_CHAR_POINTS = {
    ")": 1,
    "]": 2,
    "}": 3,
    ">": 4
}


class Puzzle(AocPuzzle[list[str], int]):
    def __init__(self):
        super().__init__(TEST_ANSWERS)


    def run_part1(self):
        illegal_chars: list[str] = []
        for line in self._input:
            stack: list[str] = []
            for char in line.strip():
                if char in CHAR_MAP:
                    stack.append(char)
                elif char != CHAR_MAP[stack.pop()]:
                    illegal_chars.append(char)
                    break

        return sum([ILLEGAL_CHAR_POINTS[char] for char in illegal_chars])


    def run_part2(self):
        scores: list[int] = []
        for line in self._input:
            stack: list[str] = []
            for char in line.strip():
                if char in CHAR_MAP:
                    stack.append(char)
                elif char != CHAR_MAP[stack.pop()]:
                    break
            else:
                stack.reverse()
                scores.append(reduce(lambda score, char: score * 5 + COMPLETION_CHAR_POINTS[CHAR_MAP[char]], stack, 0))

        return statistics.median(scores)
